### PR TITLE
Audit allowed and blocked licenses

### PR DESCRIPTION
**Pull request by ranisalt**
Added a command to audit (or verify) the list of licenses in order to warn if certain licenses are used. Usage example is `yarn licenses audit --blocked GPL-3.0 --blocked SSPL-1.0`

Currently there are a few decisions to be made that I'd like to discuss:

- There are two flags `--allowed` and `--blocked` which are exclusive, and can be set multiple times. However, it could also be a separate config file (what [tophat/yarn-plugin-licenses](https://github.com/tophat/yarn-plugin-licenses?tab=readme-ov-file#usage) does) or a field in package.json
- The command simply outputs the licenses that violate the flags. Should it have an informative message when all licenses are allowed? Should it exit with error code != 0?
- Should the command be named `audit`, `verify`, or maybe a flag in the `list` command i.e. `yarn licenses list --verify`?

### CODE REVIEW

**Comment by mhassan1** [comment]
Thanks for the PR!

> * There are two flags `--allowed` and `--blocked` which are exclusive, and can be set multiple times. However, it could also be a separate config file (what [tophat/yarn-plugin-licenses](https://github.com/tophat/yarn-plugin-licenses?tab=readme-ov-file#usage) does) or a field in package.json

Those flags make sense to me. I'd like to avoid configuration files, since it adds another layer of complexity to the plugin.

> * The command simply outputs the licenses that violate the flags. Should it have an informative message when all licenses are allowed? Should it exit with error code != 0?

I would think one use case would be to audit that no licensing rules are being violated in CI, so a non-zero exit code if the report is non-empty makes sense to me. If the report is empty, I think it should exit zero and write an empty response.

> * Should the command be named `audit`, `verify`, or maybe a flag in the `list` command i.e. `yarn licenses list --verify`?

I don't think it should be part of `list`, since it's confusing to me that `list --allowed` would list violations of that rule. I think you're correct that it should be a new command, and I think `audit` is good.

File: src/utils.ts [hunk 1]
```diff
@@ -32,17 +32,18 @@ export const pluginRootDir: PortablePath =
...
   project: Project,
   json: boolean,
   recursive: boolean,
   production: boolean,
   excludeMetadata: boolean
-): Promise<treeUtils.TreeNode> => {
-  const rootChildren: treeUtils.TreeMap = {}
-  const root: treeUtils.TreeNode = { children: rootChildren }
+) => {
+  const root = {
```
Comment (mhassan1) [inline src/utils.ts:44]: is this a functional change? or a style thing?

File: src/commands/audit.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  })
+
+  blockedLicenses = Option.Array(`--blocked`, [], {
+    description: `List of blocked licenses`
+  })
+
+  static usage: Usage = Command.Usage({
+    description: `display the licenses for all packages in the project`,
+    details: `
+      This command prints the license information for packages in the project. By default, only direct dependencies are audited.
```
Comment (mhassan1) [inline src/commands/audit.ts:40]: this line in the `details` looks incorrect

File: src/utils.ts (hunk 1 above)
Comment (ranisalt) [reply src/utils.ts:44]: Type asserting it as `TreeMap` meant that children could be another `TreeMap`, an array of `TreeNode` or `false`, which created some ugly code to check later on. However, I just quickly put that and will improve this code, probably with `satisfies`

File: src/commands/audit.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  allowedLicenses = Option.Array(`--allowed`, [], {
+    description: `List of allowed licenses`
+  })
+
+  blockedLicenses = Option.Array(`--blocked`, [], {
+    description: `List of blocked licenses`
+  })
+
+  static usage: Usage = Command.Usage({
+    description: `display the licenses for all packages in the project`,
```
[the remark this answers is not in the document]
Comment (ranisalt) [reply src/commands/audit.ts:38]: It is, I copy-pasted from `list` and forgot to change 😆

**Comment by ranisalt** [comment]
I believe I have addressed most of the concerns, I just need to implement a flag to exit with error code and then I'll mark it as ready

File: src/commands/audit.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  })
+
+  blockedLicenses = Option.Array(`--blocked`, [], {
+    description: `List of blocked licenses`
+  })
+
+  static usage: Usage = Command.Usage({
+    description: `audits the licenses for all packages in the project`,
+    details: `
+      This command audits and prints the license information for packages in the project according to rules set. By default, only direct dependencies are audited.
```
Comment (mhassan1) [inline src/commands/audit.ts:40]: nit: `according to rules set` => `that violate the rules set`

File: src/utils.ts [hunk 5]
```diff
@@ -32,17 +32,17 @@ export const pluginRootDir: PortablePath =
...
- * @returns {treeUtils.TreeNode} Root tree node
+ * @returns Root tree node
  */
 export const getTree = async (
   project: Project,
   json: boolean,
   recursive: boolean,
   production: boolean,
   excludeMetadata: boolean
-): Promise<treeUtils.TreeNode> => {
```
[the remark this answers is not in the document]
Comment (mhassan1) [reply src/utils.ts:43]: is there any way to keep the explicit return type? if need be, we can create a local type.

File: src/commands/audit.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
+import { WorkspaceRequiredError } from '@yarnpkg/cli'
+import { CommandContext, Configuration, Project, treeUtils } from '@yarnpkg/core'
+import { Command, Usage, Option } from 'clipanion'
+import { getTree, focusWorkspaces } from '../utils'
+
+export class LicensesAuditCommand extends Command<CommandContext> {
+  static paths = [[`licenses`, `audit`]]
```
Comment (mhassan1) [inline src/commands/audit.ts:7]: let's add this new command to the `README`

File: src/utils.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ranisalt) [reply src/utils.ts:43]: Yes, I can create a new type that's more specific than the previous one

File: src/commands/audit.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+  allowedLicenses = Option.Array(`--allowed`, [], {
+    description: `List of allowed licenses`
+  })
+
+  blockedLicenses = Option.Array(`--blocked`, [], {
+    description: `List of blocked licenses`
+  })
+
+  exitCode = Option.Boolean(`--exit-code`, false, {
```
Comment (mhassan1) [inline src/commands/audit.ts:37]: what's the use case for this new flag? i think we should always exit non-zero if there are violations. if the user doesn't like that, they can ignore the non-zero exit code themselves.

Comment (ranisalt) [reply src/commands/audit.ts:37]: I found testing this to be very annoying because when the exit code is non-zero:
- jest API to check for errors is very limited, it's not possible to use `toThrowErrorMatchingSnapshot` because the error thrown is stringified to `"Command failed: yarn licenses audit ..."`
- if using `toThrowError(expect.objectContaining({ ... }))` we can't use snapshots
- for reasons I don't know, `spawnSync` (or the caught error from `execSync`) will _always_ return `null` for `status` and `stdout`
So far I didn't find a way that reliably returns the stdout and the status in a way that allows me to snapshot test the stdout.

Comment (ranisalt) [reply src/commands/audit.ts]: It seems that passing `shell: true` to `spawnSync` fixes this. The documentation on this option is a bit lacking, but it should be safe to use only for testing.

File: src/__tests__/integration/audit.test.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    ['without metadata', '--exclude-metadata'],
+    ['without metadata as json', '--json --exclude-metadata']
+  ])(`should audit %s`, (description, flags) => {
+    it.each([
+      ['allowed', '--allowed MIT'],
+      ['blocked', `--blocked Apache-2.0 --blocked UNKNOWN`]
+    ])(`with %s licenses`, async (description, licenses) => {
+      const { stdout, status } = spawnSync(`yarn licenses audit ${licenses} ${flags}`, { cwd, shell: true })
+      expect(stdout.toString()).toMatchSnapshot()
+      expect(status).toBe(1)
```
Comment (mhassan1) [inline src/__tests__/integration/audit.test.ts:33]: we should add one test case where exit code is zero

**Comment by mhassan1** [comment]
Thank you for the awesome contribution!
